callouts/java/service-callout/src/main/java/service/ServiceCallout.java

```diff
@@ -0,0 +1,405 @@
+
+            plaintextServer = ServerBuilder.forPort(plaintextPort)
+                    .addService(processor)
+                    .executor(Executors.newFixedThreadPool(serverThreadCount)) // Configurable thread pool
+                    .build()
+                    .start();
+
+            logger.info("Plaintext Server started, listening on " + plaintextPort);
+        }
+
+
+
+        // Start Health Check Server if enabled
+        if (!combinedHealthCheck) {
+            healthCheckServer.start();
+            logger.info("Health Check Server started, listening on " + healthCheckIp + ":" + healthCheckPort + " at path " + healthCheckPath);
+        }
+
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            logger.info("*** shutting down gRPC server since JVM is shutting down");
+            try {
+                ServiceCallout.this.stop();
+            } catch (InterruptedException e) {
+                e.printStackTrace(System.err);
+            }
+            logger.info("*** server shut down");
+        }));
+    }
+
+    /**
+     * Stops the gRPC server gracefully.
+     * This method will attempt to shut down the server within a 30-second timeout.
+     *
+     * @throws InterruptedException If the shutdown process is interrupted.
+     */
+    private void stop() throws InterruptedException {
+        if (server != null) {
+            server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
+        }
+
+        if (plaintextServer != null) {
+            plaintextServer.shutdown().awaitTermination(30, TimeUnit.SECONDS);
+        }
+
+        if (!combinedHealthCheck && healthCheckServer != null) {
+            healthCheckServer.stop(0); // 0 delay for immediate stop
+            logger.info("Health Check Server stopped.");
+        }
+    }
+
+    /**
+     * Blocks the main thread until the server is terminated.
+     * This is necessary because gRPC uses daemon threads by default.
+     *
+     * @throws InterruptedException If interrupted while waiting for the server to shut down.
+     */
+    public void blockUntilShutdown() throws InterruptedException {
+        if (server != null) {
+            server.awaitTermination();
+        }
+
+        if (plaintextServer != null) {
+            plaintextServer.awaitTermination();
+        }
+    }
+
+    /**
+     * Processes incoming {@link ProcessingRequest} and builds the corresponding {@link ProcessingResponse}.
+     * This method handles different types of requests (headers, body) and delegates to specific
+     * handlers for processing.
+     *
+     * @param request The request to be processed.
+     * @return The processed response.
+     */
+    public ProcessingResponse processRequest(ProcessingRequest request) {
+        ProcessingResponse.Builder builder = ProcessingResponse.newBuilder();
+
+        switch (request.getRequestCase()) {
+            case REQUEST_HEADERS:
+                onRequestHeaders(builder, request.getRequestHeaders());
```

Comment:
If this aims to be a general-purpose server, then should it also pass other parts of the ProcessingRequest into the request-type-specific handler, i.e. `metadata_context` and `attributes`? (and perhaps `observability_mode` for completeness, though that seems less important)

Reply:
Per offline discussion, the current state of not passing `metadata_context` and `attributes` is consistent with other existing callout servers, so doesn't need to be included here. We can always consider in the future whether to add them in general.